concrete-core/src/backends/core/implementation/engines/ggsw_ciphertext_discarding_encryption.rs

```diff
@@ -0,0 +1,161 @@
+};
+
+/// # Description:
+/// Implementation of [`GgswCiphertextDiscardingEncryptionEngine`] for [`CoreEngine`] that operates
+/// on 32 bits integers.
+impl GgswCiphertextDiscardingEncryptionEngine<GlweSecretKey32, Plaintext32, GgswCiphertext32>
+    for CoreEngine
+{
+    /// # Example:
+    /// ```
+    /// use concrete_commons::dispersion::Variance;
+    /// use concrete_commons::parameters::{
+    ///     DecompositionBaseLog, DecompositionLevelCount, GlweDimension, PolynomialSize,
+    /// };
+    /// use concrete_core::prelude::*;
+    /// # use std::error::Error;
+    ///
+    /// # fn main() -> Result<(), Box<dyn Error>> {
+    /// // DISCLAIMER: the parameters used here are only for test purpose, and are not secure.
+    /// let glwe_dimension = GlweDimension(2);
+    /// let polynomial_size = PolynomialSize(4);
+    /// let level = DecompositionLevelCount(1);
+    /// let base_log = DecompositionBaseLog(4);
+    /// // Here a hard-set encoding is applied (shift by 20 bits)
+    /// let input = 3_u32 << 20;
+    /// let noise = Variance(2_f64.powf(-25.));
+    ///
+    /// let mut engine = CoreEngine::new()?;
+    /// let key_1: GlweSecretKey32 = engine.create_glwe_secret_key(glwe_dimension, polynomial_size)?;
+    /// let plaintext = engine.create_plaintext(&input)?;
+    /// let mut ciphertext =
+    ///     engine.encrypt_ggsw_ciphertext(&key_1, &plaintext, noise, level, base_log)?;
+    /// // We're going to re-encrypt the input with another secret key
+    /// // For this, it is required that the second secret key uses the same GLWE dimension
+    /// // and polynomial size as the first one.
+    /// let key_2: GlweSecretKey32 = engine.create_glwe_secret_key(glwe_dimension, polynomial_size)?;
+    ///
+    /// engine.discard_encrypt_ggsw_ciphertext(&key_2, &mut ciphertext, &plaintext, noise)?;
+    /// #
+    /// assert_eq!(ciphertext.glwe_dimension(), glwe_dimension);
+    /// assert_eq!(ciphertext.polynomial_size(), polynomial_size);
+    ///
+    /// engine.destroy(ciphertext)?;
+    /// engine.destroy(plaintext)?;
+    /// engine.destroy(key_1)?;
+    /// engine.destroy(key_2)?;
+    /// #
+    /// # Ok(())
+    /// # }
+    /// ```
+    fn discard_encrypt_ggsw_ciphertext(
+        &mut self,
+        key: &GlweSecretKey32,
+        output: &mut GgswCiphertext32,
+        input: &Plaintext32,
+        noise: Variance,
+    ) -> Result<(), GgswCiphertextDiscardingEncryptionError<Self::EngineError>> {
+        GgswCiphertextDiscardingEncryptionError::perform_generic_checks(key, output)?;
+        unsafe { self.discard_encrypt_ggsw_ciphertext_unchecked(key, output, input, noise) };
+        Ok(())
+    }
+
+    unsafe fn discard_encrypt_ggsw_ciphertext_unchecked(
+        &mut self,
+        key: &GlweSecretKey32,
+        output: &mut GgswCiphertext32,
+        input: &Plaintext32,
+        noise: Variance,
+    ) {
+        key.0.encrypt_constant_ggsw(
```

Comment:
Why is it an encryption of a constant GGSW here?

Reply:
It is constant as in "the plaintext is a constant", eg a single value. We did not implement any way to encrypt a plaintext on every coefficient of each polynomial yet, since in practice GGSWs are only used to make CMUX keys.

Reply:
Thanks @orelcosseron for the clarification. I agree with Alex that we could make it explicit in the name of the engine.

Reply:
I'm resolving this conversation since it is a duplicate of another one.

Reply:
Quoting @aPere3 to have everything at the same place:
Hello,
@damienligier , for some reasons, I am not able to answer on the conversation where you mentioned me, so I answer here.

I am not sure I understand your question, but from @ilaria-zama review, I imagine it is about having the possibility to encrypt ggsw ciphertext with either a constant polynomial, or a regular one, is that right ?

If we want to do that, the proper way would be to have two engine traits:

    One GgswCiphertextConstantEncryptionEngine trait which takes a plaintext entity as input
    One GgswCiphertextEncryptionEngine trait which takes a plaintext vector entity as input

Reply:
Yes, that's what I wanted to say :)